README.md
# Criador de Postagens
## Um simples criador de postagens para meu site pessoal no GitHub!

### Descrição:

###### Este é um simples programa escrito em Python e utiliza PyQt4 para a criação rápida de posts em meu site pessoal no GitHub.

###### Decidi criar este programa para facilitar a minha vida na hora de postar algo novo em meu site pessoal, já que antes eu tinha que editar manualmente o arquivo .html do novo post.

###### O programa ainda é bem simples e utiliza algumas tags para a criação de um post simples. Pretendo adicionar novas funcionalidades no futuro, para facilitar ainda mais a criação de novos posts e adicionar novas tags e funcionalidades.

###### É OBRIGATÓRIO que o arquivo 'model.html' esteja presente na pasta principal junto ao programa, este arquivo é usado como base para a criação de um novo post.

###### O arquivo 'model.html' NÃO PODE SER ALTERADO DE MANEIRA ALGUMA! O programa verifica o MD5 do arquivo modelo e faz uma comparação com a hash que ele possui do arquivo original. Se houver mudanças o programa NÃO IRÁ INICIAR, dando um aviso de que o arquivo modelo foi modificado ou corrompido!

###### O programa também irá gerar uma template para a chamada do post dentro do arquivo criado! Basta copiar este template e colar no arquivo 'posts.html', logo abaixo do exemplo de postagem, para que possa ficar no topo da lista de postagens.

###### Caso alguém ache útil o meu programa e quiser utilizar em seu próprio website, terá de usar um outro arquivo modelo e alterar a hash definida no programa para que ele possa aceitar o novo modelo! Assim como trocar alguns templates de lugar para que o programa possa substituir bem o título, o subtítulo, a data e hora da postagem, o link da postagem e a postagem completa!

### [Verifique o CHANGELOG para maiores informações sobre novas versões!](https://raw.github.com/Wolfterro/Criador-De-Postagens/master/CHANGELOG.txt)

### Requisitos:

#### Compilando:
- Python 2.7
- PyQt4 para Python 2.7
- PyInstaller
- Microsoft Visual C++ 2010 Redistributable (Windows apenas)

##### Para compilar o programa no Windows, basta executar o arquivo 'build.bat'.
##### Para compilar o programa no Linux, basta executar o arquivo 'build.sh'.
##### **OBS**: É necessário que o PyInstaller seja reconhecido como comando interno no prompt de comando (Windows) ou no shell que estiver utilizando (Linux).

### Download:

#### ***Binário (Windows):*** https://github.com/Wolfterro/Criador-De-Postagens/releases/tag/v1.2-Windows
#### ***Código-fonte:*** https://github.com/Wolfterro/Criador-De-Postagens/archive/master.zip


CHANGELOG.txt
Criador de Postagens
====================

Um simples criador de postagens para meu site pessoal no GitHub! 
----------------------------------------------------------------

CHANGELOG:
==========

VERSÃO 1.2:
- Template para a chamada da postagem é gerada pelo programa no arquivo criado.
--- Basta copiar o template e colar no arquivo 'posts.html', logo abaixo do exemplo de postagem.

#-----------------------------------------------------------------------------------------------

VERSÃO 1.1:
- Variável com o valor da hash MD5 do arquivo modelo movida para a  classe GlobalVars.
- Adicionada opção de salvar diretamente o arquivo já aberto pelo programa.
- Adicionada opção de limpar todos os campos e fechar o arquivo já aberto pelo programa.

#-----------------------------------------------------------------------------------------------

VERSÃO 1.0:
- Versão inicial do programa.

#-----------------------------------------------------------------------------------------------


src/old/v1.1/GlobalVars.py
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.1 - Python 2.x
# Data: 26/03/2017
#===================================

class GlobalVars:
	Version = "1.1"
	ModelFileMD5 = "e5f02f183eb2dea4cfda8f97c8e39b5e"

src/old/v1.1/CriadorDePostagens.py
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.1 - Python 2.x
# Data: 26/03/2017
#===================================

from PyQt4 import QtCore, QtGui
import sys

# Imports do programa
# ===================
from WindowHandler import WindowHandler
from GlobalVars import GlobalVars

# Definindo a codificação padrão para UTF-8.
# ==========================================
reload(sys)
sys.setdefaultencoding('utf-8')

# Codificação do programa.
# ========================
try:
	_fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
	def _fromUtf8(s):
		return s

try:
	_encoding = QtGui.QApplication.UnicodeUTF8
	def _translate(context, text, disambig):
		return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
	def _translate(context, text, disambig):
		return QtGui.QApplication.translate(context, text, disambig)

# Classe principal do Programa gerado pelo Qt Designer.
# =====================================================
class Ui_MainWindow(object):
	def setupUi(self, MainWindow, Handler):
		MainWindow.setObjectName(_fromUtf8("MainWindow"))
		MainWindow.resize(700, 820)
		icon = QtGui.QIcon()
		icon.addPixmap(QtGui.QPixmap(_fromUtf8("Icon.ico")), QtGui.QIcon.Normal, QtGui.QIcon.Off)
		MainWindow.setWindowIcon(icon)
		self.centralwidget = QtGui.QWidget(MainWindow)
		self.centralwidget.setObjectName(_fromUtf8("centralwidget"))
		self.gridLayout_5 = QtGui.QGridLayout(self.centralwidget)
		self.gridLayout_5.setObjectName(_fromUtf8("gridLayout_5"))
		self.groupBox = QtGui.QGroupBox(self.centralwidget)
		self.groupBox.setObjectName(_fromUtf8("groupBox"))
		self.gridLayout_4 = QtGui.QGridLayout(self.groupBox)
		self.gridLayout_4.setObjectName(_fromUtf8("gridLayout_4"))
		self.lineEdit = QtGui.QLineEdit(self.groupBox)
		self.lineEdit.setObjectName(_fromUtf8("lineEdit"))
		self.gridLayout_4.addWidget(self.lineEdit, 0, 0, 1, 1)
		self.gridLayout_5.addWidget(self.groupBox, 0, 0, 1, 1)
		self.groupBox_2 = QtGui.QGroupBox(self.centralwidget)
		self.groupBox_2.setObjectName(_fromUtf8("groupBox_2"))
		self.gridLayout_3 = QtGui.QGridLayout(self.groupBox_2)
		self.gridLayout_3.setObjectName(_fromUtf8("gridLayout_3"))
		self.lineEdit_2 = QtGui.QLineEdit(self.groupBox_2)
		self.lineEdit_2.setObjectName(_fromUtf8("lineEdit_2"))
		self.gridLayout_3.addWidget(self.lineEdit_2, 0, 0, 1, 1)
		self.gridLayout_5.addWidget(self.groupBox_2, 1, 0, 1, 1)
		self.groupBox_3 = QtGui.QGroupBox(self.centralwidget)
		self.groupBox_3.setObjectName(_fromUtf8("groupBox_3"))
		self.gridLayout_2 = QtGui.QGridLayout(self.groupBox_3)
		self.gridLayout_2.setObjectName(_fromUtf8("gridLayout_2"))
		self.textEdit = QtGui.QTextEdit(self.groupBox_3)
		self.textEdit.setObjectName(_fromUtf8("textEdit"))
		self.gridLayout_2.addWidget(self.textEdit, 0, 0, 1, 1)
		self.gridLayout_5.addWidget(self.groupBox_3, 2, 0, 1, 1)
		self.groupBox_4 = QtGui.QGroupBox(self.centralwidget)
		self.groupBox_4.setObjectName(_fromUtf8("groupBox_4"))
		self.gridLayout = QtGui.QGridLayout(self.groupBox_4)
		self.gridLayout.setObjectName(_fromUtf8("gridLayout"))
		self.pushButton = QtGui.QPushButton(self.groupBox_4)
		self.pushButton.setStyleSheet(_fromUtf8("QPushButton {\n"
"    font-weight: bold;\n"
"}"))
		self.pushButton.setObjectName(_fromUtf8("pushButton"))
		self.gridLayout.addWidget(self.pushButton, 0, 0, 1, 1)
		self.pushButton_2 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_2.setStyleSheet(_fromUtf8("QPushButton {\n"
"    font-style: italic;\n"
"}"))
		self.pushButton_2.setObjectName(_fromUtf8("pushButton_2"))
		self.gridLayout.addWidget(self.pushButton_2, 0, 1, 1, 1)
		self.pushButton_3 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_3.setStyleSheet(_fromUtf8("QPushButton {\n"
"    text-decoration: underline;\n"
"}"))
		self.pushButton_3.setObjectName(_fromUtf8("pushButton_3"))
		self.gridLayout.addWidget(self.pushButton_3, 0, 2, 1, 2)
		self.pushButton_4 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_4.setStyleSheet(_fromUtf8(""))
		self.pushButton_4.setObjectName(_fromUtf8("pushButton_4"))
		self.gridLayout.addWidget(self.pushButton_4, 0, 4, 1, 2)
		self.pushButton_5 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_5.setObjectName(_fromUtf8("pushButton_5"))
		self.gridLayout.addWidget(self.pushButton_5, 0, 6, 1, 1)
		self.pushButton_6 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_6.setObjectName(_fromUtf8("pushButton_6"))
		self.gridLayout.addWidget(self.pushButton_6, 0, 7, 1, 2)
		self.pushButton_7 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_7.setObjectName(_fromUtf8("pushButton_7"))
		self.gridLayout.addWidget(self.pushButton_7, 0, 9, 1, 1)
		self.pushButton_14 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_14.setObjectName(_fromUtf8("pushButton_14"))
		self.gridLayout.addWidget(self.pushButton_14, 0, 10, 1, 1)
		self.pushButton_8 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_8.setObjectName(_fromUtf8("pushButton_8"))
		self.gridLayout.addWidget(self.pushButton_8, 1, 0, 1, 1)
		self.pushButton_9 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_9.setObjectName(_fromUtf8("pushButton_9"))
		self.gridLayout.addWidget(self.pushButton_9, 1, 1, 1, 2)
		self.pushButton_10 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_10.setObjectName(_fromUtf8("pushButton_10"))
		self.gridLayout.addWidget(self.pushButton_10, 1, 3, 1, 2)
		self.pushButton_11 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_11.setObjectName(_fromUtf8("pushButton_11"))
		self.gridLayout.addWidget(self.pushButton_11, 1, 5, 1, 2)
		self.pushButton_12 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_12.setObjectName(_fromUtf8("pushButton_12"))
		self.gridLayout.addWidget(self.pushButton_12, 1, 7, 1, 1)
		self.pushButton_13 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_13.setObjectName(_fromUtf8("pushButton_13"))
		self.gridLayout.addWidget(self.pushButton_13, 1, 8, 1, 2)
		self.pushButton_15 = QtGui.QPushButton(self.groupBox_4)
		self.pushButton_15.setObjectName(_fromUtf8("pushButton_15"))
		self.gridLayout.addWidget(self.pushButton_15, 1, 10, 1, 1)
		self.gridLayout_5.addWidget(self.groupBox_4, 3, 0, 1, 1)
		MainWindow.setCentralWidget(self.centralwidget)
		self.menubar = QtGui.QMenuBar(MainWindow)
		self.menubar.setGeometry(QtCore.QRect(0, 0, 691, 20))
		self.menubar.setObjectName(_fromUtf8("menubar"))
		self.menuArquivo = QtGui.QMenu(self.menubar)
		self.menuArquivo.setObjectName(_fromUtf8("menuArquivo"))
		self.menuFormatar = QtGui.QMenu(self.menubar)
		self.menuFormatar.setObjectName(_fromUtf8("menuArquivo"))
		MainWindow.setMenuBar(self.menubar)
		self.statusbar = QtGui.QStatusBar(MainWindow)
		self.statusbar.setObjectName(_fromUtf8("statusbar"))
		MainWindow.setStatusBar(self.statusbar)
		self.actionSalvar = QtGui.QAction(MainWindow)
		self.actionSalvar.setObjectName(_fromUtf8("actionSalvar"))
		self.actionNovo = QtGui.QAction(MainWindow)
		self.actionNovo.setObjectName(_fromUtf8("actionNovo"))
		self.actionSalvar_Como = QtGui.QAction(MainWindow)
		self.actionSalvar_Como.setObjectName(_fromUtf8("actionSalvar_Como"))
		self.actionSair = QtGui.QAction(MainWindow)
		self.actionSair.setObjectName(_fromUtf8("actionSair"))
		self.actionFonte = QtGui.QAction(MainWindow)
		self.actionFonte.setObjectName(_fromUtf8("actionFonte"))
		self.menuArquivo.addAction(self.actionNovo)
		self.menuArquivo.addAction(self.actionSalvar)
		self.menuArquivo.addAction(self.actionSalvar_Como)
		self.menuArquivo.addAction(self.actionSair)
		self.menuFormatar.addAction(self.actionFonte)
		self.menubar.addAction(self.menuArquivo.menuAction())
		self.menubar.addAction(self.menuFormatar.menuAction())

		# Adicionando evento 'clicked.connect' aos botões da janela
		# =========================================================
		self.pushButton.clicked.connect(lambda: Handler.InsertTag(u"<b></b>", True))
		self.pushButton_2.clicked.connect(lambda: Handler.InsertTag(u"<i></i>", True))
		self.pushButton_3.clicked.connect(lambda: Handler.InsertTag(u"<u></u>", True))
		self.pushButton_4.clicked.connect(lambda: Handler.InsertTag(u"<del></del>", True))
		self.pushButton_5.clicked.connect(lambda: Handler.InsertTag(u"<img class=\"img-responsive\" src=\"INSIRA O CAMINHO DA IMAGEM AQUI\" alt=\"NOME DA IMAGEM\"></img>", True))
		self.pushButton_6.clicked.connect(lambda: Handler.InsertTag(u"<a href=\"INSIRA O LINK AQUI\" target=\"_blank\"></a>", True))
		self.pushButton_7.clicked.connect(lambda: Handler.InsertTag(u"<p></p>", True))
		self.pushButton_8.clicked.connect(lambda: Handler.InsertTag(u"<h1></h1>", True))
		self.pushButton_9.clicked.connect(lambda: Handler.InsertTag(u"<h2></h2>", True))
		self.pushButton_10.clicked.connect(lambda: Handler.InsertTag(u"<h3></h3>", True))
		self.pushButton_11.clicked.connect(lambda: Handler.InsertTag(u"<center></center>", False))
		self.pushButton_12.clicked.connect(lambda: Handler.InsertTag(u"<video><source src=\"INSIRA O CAMINHO DO VÍDEO AQUI\" type=\"video/mp4\"></video>", True))
		self.pushButton_13.clicked.connect(lambda: Handler.InsertTag(u"<audio><source src=\"INSIRA O CAMINHO DO ÁUDIO AQUI\" type=\"audio/mpeg\"></audio>", True))
		self.pushButton_14.clicked.connect(lambda: Handler.InsertTag(u"<br>", True))
		self.pushButton_15.clicked.connect(lambda: Handler.InsertTag(u"<hr>", True))

		# Adicionando evento 'triggered.connect' aos menus da janela
		# ==========================================================
		self.actionNovo.triggered.connect(Handler.Clear)
		self.actionSair.triggered.connect(Handler.ExitProgram)
		self.actionSalvar.triggered.connect(Handler.Save)
		self.actionSalvar_Como.triggered.connect(Handler.SaveAs)
		self.actionFonte.triggered.connect(Handler.ChangeFont)

		self.retranslateUi(MainWindow)
		QtCore.QMetaObject.connectSlotsByName(MainWindow)

	def retranslateUi(self, MainWindow):
		MainWindow.setWindowTitle(_translate("MainWindow", "Criador de Postagens - v%s" % (GlobalVars.Version), None))
		self.groupBox.setTitle(_translate("MainWindow", "Título", None))
		self.groupBox_2.setTitle(_translate("MainWindow", "Subtítulo", None))
		self.groupBox_3.setTitle(_translate("MainWindow", "Postagem", None))
		self.groupBox_4.setTitle(_translate("MainWindow", "Ferramentas de Postagem", None))
		self.pushButton.setText(_translate("MainWindow", "B", None))
		self.pushButton_2.setText(_translate("MainWindow", "i", None))
		self.pushButton_3.setText(_translate("MainWindow", "u", None))
		self.pushButton_4.setText(_translate("MainWindow", "<del>", None))
		self.pushButton_5.setText(_translate("MainWindow", "<img>", None))
		self.pushButton_6.setText(_translate("MainWindow", "<a>", None))
		self.pushButton_7.setText(_translate("MainWindow", "<p>", None))
		self.pushButton_14.setText(_translate("MainWindow", "<br>", None))
		self.pushButton_8.setText(_translate("MainWindow", "<h1>", None))
		self.pushButton_9.setText(_translate("MainWindow", "<h2>", None))
		self.pushButton_10.setText(_translate("MainWindow", "<h3>", None))
		self.pushButton_11.setText(_translate("MainWindow", "<center>", None))
		self.pushButton_12.setText(_translate("MainWindow", "<video>", None))
		self.pushButton_13.setText(_translate("MainWindow", "<audio>", None))
		self.pushButton_15.setText(_translate("MainWindow", "<hr>", None))
		self.menuArquivo.setTitle(_translate("MainWindow", "Arquivo", None))
		self.menuFormatar.setTitle(_translate("MainWindow", "Formatar", None))
		self.actionNovo.setText(_translate("MainWindow", "Novo", None))
		self.actionSalvar.setText(_translate("MainWindow", "Salvar", None))
		self.actionSalvar_Como.setText(_translate("MainWindow", "Salvar Como...", None))
		self.actionSair.setText(_translate("MainWindow", "Sair", None))
		self.actionFonte.setText(_translate("MainWindow", "Fonte...", None))

if __name__ == "__main__":
	app = QtGui.QApplication(sys.argv)
	MainWindow = QtGui.QMainWindow()
	ui = Ui_MainWindow()
	
	# Os métodos do programa serão definidos pelo Handler
	# ---------------------------------------------------
	Handler = WindowHandler(ui, MainWindow)

	# Definindo locale do programa
	# ----------------------------
	translator = QtCore.QTranslator()
	locale = QtCore.QLocale.system().name()
	translator.load('qt_%s' % locale, 
		QtCore.QLibraryInfo.location(QtCore.QLibraryInfo.TranslationsPath))
	app.installTranslator(translator)
	
	ui.setupUi(MainWindow, Handler)
	MainWindow.show()
	sys.exit(app.exec_())

src/old/v1.1/WindowHandler.py
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.1 - Python 2.x
# Data: 26/03/2017
#===================================

from __future__ import print_function
from PyQt4 import QtCore, QtGui
from os.path import expanduser

import os
import sys
import cgi
import ctypes
import hashlib
import datetime
import platform

# Imports do programa
# ===================
from GlobalVars import GlobalVars

# Definindo a codificação padrão para UTF-8.
# ==========================================
reload(sys)
sys.setdefaultencoding('utf-8')

# Codificação do programa.
# ========================
try:
	_fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
	def _fromUtf8(s):
		return s

try:
	_encoding = QtGui.QApplication.UnicodeUTF8
	def _translate(context, text, disambig):
		return QtGui.QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
	def _translate(context, text, disambig):
		return QtGui.QApplication.translate(context, text, disambig)

# Determinando a pasta 'home' do usuário.
# !!! NÃO-IMPLEMENTADO AINDA !!!
# =======================================
if platform.system() == "Windows":
	buf = ctypes.create_unicode_buffer(1024)
	ctypes.windll.kernel32.GetEnvironmentVariableW(u"USERPROFILE", buf, 1024)
	home_dir = buf.value
else:
	home_dir = expanduser("~")

# Classe de métodos da janela principal do programa
# =================================================
class WindowHandler(object):
	# Obtendo o objeto da janela principal do programa
	# ================================================
	def __init__(self, ui, MainWindow):
		self.ui = ui
		self.MainWindow = MainWindow
		self.Filename = ""
		self.CheckModelFile()

	# Adicionando tags ao texto principal
	# ===================================
	def InsertTag(self, Tag, isNewLine):
		if isNewLine:
			self.ui.textEdit.insertPlainText("%s\n" % (Tag))
		else:
			self.ui.textEdit.insertPlainText(Tag)

	# Limpando todos os campos e zerando nome de arquivo para um novo post
	# ====================================================================
	def Clear(self):
		self.ui.textEdit.clear()
		self.ui.lineEdit.clear()
		self.ui.lineEdit_2.clear()
		self.Filename = ""
		self.ShowOpenedFilename(False)

	# Gerando texto automático para inserir como identificador no arquivo .html
	# =========================================================================
	def GenerateAutoText(self):
		return u"Página HTML gerada pelo Criador de Postagens - v%s -#-#- Timestamp: %s - %s" % (GlobalVars.Version,
			datetime.datetime.now().strftime("%d/%m/%Y"),
			datetime.datetime.now().strftime("%H:%M:%S"))

	# Gerando data formatada para inserir no arquivo .html
	# ====================================================
	def GenerateFormattedData(self):
		return "%s" % (datetime.datetime.now().strftime("%d/%m/%Y"))

	# Mostrando o arquivo aberto no título da janela
	# ==============================================
	def ShowOpenedFilename(self, isOpened):
		if isOpened:
			self.MainWindow.setWindowTitle(_translate("MainWindow", "Criador de Postagens - v%s - %s" % (GlobalVars.Version,
				self.Filename), None))
		else:
			self.MainWindow.setWindowTitle(_translate("MainWindow", "Criador de Postagens - v%s" % (GlobalVars.Version), None))

	# Salvando em um novo arquivo .html
	# =================================
	def SaveAs(self):
		Title, SubTitle, GeneratedText, FormattedData, MainPost = self.GetValues()
		
		self.ChosenFilename = QtGui.QFileDialog.getSaveFileName(None, "Salvar como", "", u"Página da Web (*.html)")

		if self.ChosenFilename != "":
			self.Filename = self.ChosenFilename
			self.SaveMethod(Title, SubTitle, GeneratedText, FormattedData, MainPost)
			self.ShowOpenedFilename(True)

	# Salvando arquivo .html aberto no programa
	# =========================================
	def Save(self):
		if self.Filename != "":
			Title, SubTitle, GeneratedText, FormattedData, MainPost = self.GetValues()
			self.SaveMethod(Title, SubTitle, GeneratedText, FormattedData, MainPost)
		else:
			self.SaveAs()

	# Método comum para salvamento de arquivo
	# =======================================
	def SaveMethod(self, Title, SubTitle, GeneratedText, FormattedData, MainPost):
		FileModel = open("model.html", "r")
		ModelText = FileModel.read()
		FileModel.close()

		ModelText = ModelText.replace("[####TÍTULO####]", Title)				# Inserindo título
		ModelText = ModelText.replace("[####TEXTO-GERADO####]", GeneratedText)	# Inserindo texto gerado
		ModelText = ModelText.replace("[####DATAFORMATADA####]", FormattedData)	# Inserindo data formatada
		ModelText = ModelText.replace("[####SUBTÍTULO####]", SubTitle)			# Inserindo subtítulo
		ModelText = ModelText.replace("[####POSTAGEM-PRINCIPAL####]", MainPost)	# Inserindo post principal

		SavedFile = open(unicode(self.Filename), "w")
		SavedFile.write(ModelText)
		SavedFile.close()

	# Resgatando os valores do programa
	# =================================
	def GetValues(self):
		Title = cgi.escape(self.ui.lineEdit.text())
		SubTitle = cgi.escape(self.ui.lineEdit_2.text())
		GeneratedText = self.GenerateAutoText()
		FormattedData = self.GenerateFormattedData()
		MainPost = self.ui.textEdit.toPlainText().replace("\n", "\n\t\t\t\t\t")

		return [Title, SubTitle, GeneratedText, FormattedData, MainPost]

	# Saindo do programa
	# ==================
	def ExitProgram(self):
		sys.exit(0)

	# Diálogo para alterar a fonte do campo de postagem (self.ui.textEdit)
	# ====================================================================
	def ChangeFont(self):
		FontDialog = QtGui.QFontDialog()
		FontDialog.setWindowIcon(QtGui.QIcon("Icon.ico"))
		SelectedFont, isOK = FontDialog.getFont(self.ui.textEdit)

		if isOK:
			self.ui.textEdit.setFont(SelectedFont)

	# Verificando o arquivo modelo para criar a postagem
	# ==================================================
	def CheckModelFile(self):
		if os.path.isfile("model.html"):
			fileMD5 = hashlib.md5(open("model.html", 'rb').read()).hexdigest()
			
			if fileMD5 != GlobalVars.ModelFileMD5:
				self.ShowMessageBox(u"Erro!", 
					QtGui.QMessageBox.Critical, 
					u"Erro de checksum MD5 no arquivo modelo!", 
					u"O programa não poderá funcionar se o arquivo modelo 'model.html' estiver alterado ou corrompido!",
					QtGui.QMessageBox.Ok,
					5)
		else:
			self.ShowMessageBox(u"Erro!", 
					QtGui.QMessageBox.Critical, 
					u"Arquivo modelo não existe!", 
					u"O programa não poderá funcionar se o arquivo modelo 'model.html' não estiver presente!",
					QtGui.QMessageBox.Ok,
					4)

	# Mostrando uma janela de aviso para o usuário apenas com o botão OK.
	# Caso o valor de fechamento seja diferente de zero, o programa irá encerrar!
	# ===========================================================================
	def ShowMessageBox(self, WindowTitle, IconType, Text, InfoText, CloseType, CloseVal):
		msg = QtGui.QMessageBox()
		msg.setIcon(IconType)
		msg.setWindowIcon(QtGui.QIcon("Icon.ico"))

		msg.setText(Text)
		msg.setInformativeText(InfoText)
		msg.setWindowTitle(WindowTitle)
		msg.setStandardButtons(CloseType)
		msg.exec_()
		
		if CloseVal != 0:
			sys.exit(CloseVal)